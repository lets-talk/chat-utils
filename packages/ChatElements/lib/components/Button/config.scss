@import "../../common/styles/variables";
@import "../../common/styles/themes/themify";

$button-height: 40px;
$button-border-radius: 20px;
$button-box-shadow: none;
