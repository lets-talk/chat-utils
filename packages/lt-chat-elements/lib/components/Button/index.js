import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './index.scss';

const DELAY = 300;
const LOADER_DELAY = 900;

class Button extends Component {
  constructor(props) {
    super(props);
    this.state = {
      disabled: false,
      showLoading: false,
    };
  }

  onClick(event) {
    const { clickHandler } = this.props;

    event.preventDefault();
    event.stopPropagation();

    this.setState({ disabled: true });

    const handler = clickHandler();
    if (handler && handler instanceof Promise) {
      const loader = setTimeout(() => this.setState({ showLoading: true }), LOADER_DELAY);
      handler.then(() => {
        clearTimeout(loader);

        setTimeout(() => this.setState({
          disabled: false,
          showLoading: false,
        }), DELAY);
      });
    } else {
      this.setState({ disabled: false });
    }
  }

  render() {
    const {
      type, className, disabled, value,
    } = this.props;

    return (
      <div className="letstalk-button-container">
        <input
          type={type}
          className={classNames('letstalk-button', className)}
          value={this.state.showLoading === true ? 'Processing...' : value}
          disabled={disabled === true ? disabled : this.state.disabled}
          onClick={(event) => this.onClick(event)}
        />
      </div>
    );
  }
}

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'submit', 'reset']).isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  value: PropTypes.string.isRequired,
  clickHandler: PropTypes.func.isRequired,
};

export default Button;
