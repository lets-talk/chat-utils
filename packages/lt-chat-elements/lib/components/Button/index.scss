@import './config';

.letstalk-button-container {
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 10px;
}

.letstalk-button {
  position: relative;
  @include themify($themes) {
    color: themed('color-text');
    background-color: themed('color-background-primary');
  }
  height: $button-height;
  border-radius: $button-border-radius;
  box-shadow: $button-box-shadow;
  padding: 10px 25px;
  float: left;
}
