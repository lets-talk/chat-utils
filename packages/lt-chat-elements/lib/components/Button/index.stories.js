import React from 'react';
// Storybook stuff
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, color, select, text } from '@storybook/addon-knobs';

import { WrapWithTheme, defaultTheme, themeOptions } from '../../utils/stories';
// Component to show on storybook
import { Button } from '../../../lib';

const stories = storiesOf('Button', module);
stories.addDecorator(withKnobs);

stories.addWithInfo(
  'Basic Usage - Button',
  'This is the basic usage of a Button.',
  () => (
    <div
      style={{
        position: 'fixed', bottom: 0, right: '10px', width: 'calc(100% - 20px)', height: '100%',
      }}
    >
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <Button
          type={select('Type', { button: 'button', submit: 'submit', reset: 'reset' }, 'button')}
          color={color('Color', '#7DBC37')}
          borderRadius={text('borderRadius', '5px')}
          className={text('className', '')}
          disabled={boolean('disabled', false)}
          value={text('value', 'Identificarse')}
          clickHandler={action('clickHandler')}
        />
      </WrapWithTheme>
    </div>
  )
);
