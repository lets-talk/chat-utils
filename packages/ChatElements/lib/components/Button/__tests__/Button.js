import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '../index';

const sleep = (time) => new Promise((resolve) => setTimeout(resolve, time));

describe('Button component', () => {
  it('should render with minimal props', () => {
    const component = shallow(<Button type="submit" value="submit" clickHandler={(e) => e} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with data', () => {
    const props = {
      type: 'button',
      className: 'my-custom-button-class',
      disabled: true,
      value: 'custom',
      clickHandler: () => null,
    };

    const component = shallow(<Button {...props} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  describe('Button with and without clickHandler', () => {
    it('Test clickHandler event without Promise', () => {
      const mockCallBack = jest.fn();
      const fakeEventObject = { preventDefault() {}, stopPropagation() {} };

      const props = {
        type: 'button',
        className: 'my-custom-button-class',
        disabled: false,
        value: 'custom',
      };

      const button = shallow((<Button clickHandler={mockCallBack} {...props} />));
      button.find('.letstalk-button').simulate('click', fakeEventObject);
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });
  });

  it('Test clickHandler event with Promise', () => {
    const mockClickHandler = jest.fn(() => Promise.resolve({}));
    const fakeEventObject = { preventDefault() {}, stopPropagation() {} };

    const props = {
      type: 'button',
      className: 'my-custom-button-class',
      disabled: false,
      value: 'custom',
    };

    const button = shallow((<Button clickHandler={mockClickHandler} {...props} />));
    button.find('.letstalk-button').simulate('click', fakeEventObject);
    expect(mockClickHandler.mock.calls.length).toEqual(1);
  });
});
