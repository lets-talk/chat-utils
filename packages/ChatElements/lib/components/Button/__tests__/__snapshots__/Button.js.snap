// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render with data 1`] = `
<div
  className="letstalk-button-container"
>
  <input
    className="letstalk-button my-custom-button-class"
    disabled={true}
    onClick={[Function]}
    type="button"
    value="custom"
  />
</div>
`;

exports[`Button component should render with minimal props 1`] = `
<div
  className="letstalk-button-container"
>
  <input
    className="letstalk-button"
    disabled={false}
    onClick={[Function]}
    type="submit"
    value="submit"
  />
</div>
`;
